solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/outbound/JCSMPOutboundMessageHandler.java

```diff
@@ -96,52 +101,106 @@ public void handleMessage(Message<?> message) throws MessagingException {
 					String.format("Unable to parse header %s", SolaceBinderHeaders.CONFIRM_CORRELATION), e);
 		}
 
-		XMLMessage xmlMessage = xmlMessageMapper.map(message, properties.getExtension().getHeaderExclusions(),
-				properties.getExtension().isNonserializableHeaderConvertToString());
-		correlationKey.setRawMessage(xmlMessage);
-		xmlMessage.setCorrelationKey(correlationKey);
+		List<XMLMessage> smfMessages;
+		List<Destination> dynamicDestinations;
+		if (message.getHeaders().containsKey(SolaceBinderHeaders.BATCHED_HEADERS)) {
+			LOGGER.debug("Detected header {}, handling as batched message (Message<List<?>>) <message handler ID: {}>",
+					SolaceBinderHeaders.BATCHED_HEADERS, id);
+			smfMessages = xmlMessageMapper.mapBatchMessage(
+					message,
+					properties.getExtension().getHeaderExclusions(),
+					properties.getExtension().isNonserializableHeaderConvertToString());
+
+			BatchProxyCorrelationKey batchProxyCorrelationKey = transactedSession == null ?
+					new BatchProxyCorrelationKey(correlationKey, smfMessages.size()) : null;
+			smfMessages.forEach(smfMessage -> smfMessage.setCorrelationKey(
+					Objects.requireNonNullElse(batchProxyCorrelationKey, correlationKey)));
+
+			// after successfully running xmlMessageMapper.mapBatchMessage(),
+			// SolaceBinderHeaders.BATCHED_HEADERS is verified to be well-formed
 
-		LOGGER.debug("Publishing message to destination [ {}:{} ]",
-				targetDestination instanceof Topic ? "TOPIC" : "QUEUE", targetDestination);
+			@SuppressWarnings("unchecked")
+			List<Map<String, Object>> batchedHeaders = (List<Map<String, Object>>) message.getHeaders()
+					.getOrDefault(SolaceBinderHeaders.BATCHED_HEADERS,
+							Collections.nCopies(smfMessages.size(), Collections.emptyMap()));
+
+			dynamicDestinations = batchedHeaders.stream()
+					.map(h -> getDynamicDestination(h, correlationKey))
+					.toList();
+		} else {
+			XMLMessage smfMessage = xmlMessageMapper.map(
+					message,
+					properties.getExtension().getHeaderExclusions(),
+					properties.getExtension().isNonserializableHeaderConvertToString());
+			smfMessage.setCorrelationKey(correlationKey);
+			smfMessages = List.of(smfMessage);
+			dynamicDestinations = Collections.singletonList(getDynamicDestination(message.getHeaders(), correlationKey));
+		}
+
+		correlationKey.setRawMessages(smfMessages);
 
 		try {
-			producer.send(xmlMessage, targetDestination);
+			for (int i = 0; i < smfMessages.size(); i++) {
+				XMLMessage smfMessage = smfMessages.get(i);
```

Comment:
I believe that a potential improvement would be to set [ackImmediately()](https://docs.solace.com/API-Developer-Online-Ref-Documentation/java/com/solacesystems/jcsmp/XMLMessage.html#setAckImmediately-boolean-) on the last message of a batch. 

Without the above, thresholds have to be hit before acknowledgments are sent, see [Ack Immediately](https://docs.solace.com/API/API-Developer-Guide/Creating-Messages-1.htm) docs. 

In batch mode, it could be beneficial to short-circuit these thresholds so that acks are sent faster (exactly when needed).

Reply:
Fixed. I set `ackImmediately` to `true` on the last message in non-transacted batches.